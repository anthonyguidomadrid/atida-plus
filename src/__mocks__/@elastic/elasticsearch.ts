export const get = jest.fn()

export const mget = jest.fn()

export const Client = jest.fn().mockImplementation(() => ({
  get,
  mget,
  on: jest.fn()
}))
