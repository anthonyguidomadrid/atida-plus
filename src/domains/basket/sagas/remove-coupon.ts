import type { SagaIterator } from 'redux-saga'
import {
  all,
  call,
  getContext,
  put,
  select,
  takeEvery
} from 'redux-saga/effects'
import { createClient, logger } from '~helpers'
import {
  removeCouponTrigger,
  removeCouponFailure,
  removeCouponSuccess,
  removeCouponFulfill,
  removeCouponRequest
} from '../slices'
import { AxiosResponse } from 'axios'
import { selectId } from '~domains/basket'
import type { PayloadAction } from '@reduxjs/toolkit'
import { getErrorMessage } from '~helpers/error'

function* removeCoupon({ payload }: PayloadAction<string>) {
  try {
    const locale: string = yield getContext('locale')
    const client = createClient({
      locale,
      addAnonymousCustomerUniqueId: true
    })
    const cartId: string = yield select(selectId)

    yield put(removeCouponRequest())

    const apiResponse: AxiosResponse = yield call(
      client.delete,
      '/api/basket/remove-coupon',
      {
        data: {
          cartId,
          coupon: payload
        }
      }
    )
    yield put(removeCouponSuccess(apiResponse.data))
  } catch (error) {
    const errorMessages = getErrorMessage(error)
    logger.error(errorMessages)
    if (Array.isArray(errorMessages)) {
      yield put(
        removeCouponFailure({
          message: errorMessages[0]
        })
      )
    } else {
      yield put(
        removeCouponFailure({
          message: errorMessages
        })
      )
    }
  } finally {
    yield put(removeCouponFulfill())
  }
}

export function* removeCouponSaga(): SagaIterator {
  yield all([takeEvery(removeCouponTrigger, removeCoupon)])
}
