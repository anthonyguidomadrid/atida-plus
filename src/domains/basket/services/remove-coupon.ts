import getConfig from 'next/config'
import { createClient } from '~helpers'
import { normalizeBasket } from '~domains/basket'
import {
  Basket,
  SprykerBasketItem,
  SprykerBaskets
} from '~domains/basket/types'
import { getBasketApiConfig } from '~domains/basket/helpers'
import { fetchProductsForBasket } from './fetch-products-for-basket'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const removeCoupon = async (
  locale: string,
  cartId: string | string[],
  coupon: string | string[],
  user: { anonymousCustomerUniqueId?: string; token?: string }
): Promise<Basket> => {
  const { serverRuntimeConfig } = getConfig()
  const client = createClient({
    locale,
    options: {
      baseURL: serverRuntimeConfig.ecommerceHost[locale],
      headers: user.token
        ? {
            Authorization: `Bearer ${user.token}`
          }
        : undefined
    },
    addAnonymousCustomerUniqueId: !!user.token
      ? undefined
      : user.anonymousCustomerUniqueId
  })

  const apiConfig = getBasketApiConfig(user.token)
  const deleteRequestUrl = !!user.token
    ? `/carts/${cartId}/vouchers/${coupon}`
    : `/guest-carts/${cartId}/vouchers/${coupon}`

  await client.delete(deleteRequestUrl)

  const response = await client.get<SprykerBaskets>(apiConfig.cartUrl, {
    params: {
      include: `${apiConfig.cartItemsType},promotional-items`
    }
  })
  return fetchProductsForBasket(
    locale,
    normalizeBasket({
      data: response?.data?.data?.[0],
      included: response?.data?.data?.[0]?.relationships?.[
        apiConfig.cartItemsType
      ]?.data
        .map(({ id }) =>
          response?.data?.included?.find(
            includedItem => id === includedItem?.id
          )
        )
        .filter((item): item is SprykerBasketItem => !!item)
        .concat(
          ...response?.data?.included?.filter(
            ({ type }) => type === 'promotional-items'
          )
        )
    })
  )
}
