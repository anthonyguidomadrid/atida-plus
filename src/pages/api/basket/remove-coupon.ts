import type { NextApiRequest, NextApiResponse } from 'next'
import { getGuestJWTName, getJWTName } from '~domains/account'
import { removeCoupon } from '~domains/basket/services/remove-coupon'
import { getDefaultLocale } from '~domains/translated-routes'
import { handleApiError, isApiError, handleUnknownError } from '~helpers/error'

export default async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  try {
    const locale = req.headers['accept-language'] ?? getDefaultLocale()
    const token = req.cookies[getJWTName()] ?? req.cookies[getGuestJWTName()]
    const anonymousCustomerUniqueId = req.headers[
      'x-anonymous-customer-unique-id'
    ] as string
    const coupon = req.body.coupon
    const cartId = req.body.cartId
    const result = await removeCoupon(locale, cartId, coupon, {
      anonymousCustomerUniqueId,
      token
    })
    res.statusCode = 200
    res.json(result)
  } catch (error) {
    if (isApiError(error)) {
      handleApiError(res, error, 'spryker-cart.error.code_cant_be_removed')
    } else {
      handleUnknownError(
        req,
        res,
        error,
        'spryker-cart.error.code_cant_be_removed'
      )
    }
  }
}
