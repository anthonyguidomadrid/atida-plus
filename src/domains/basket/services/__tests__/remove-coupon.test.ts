/**
 * @jest-environment node
 */
import axios from 'axios'
// @ts-ignore
import { mget } from '@elastic/elasticsearch'
import {
  authenticatedSprykerBasket,
  elasticsearchProducts,
  sprykerBasket
} from '~domains/basket/__mocks__/basket'
import { createClient } from '~helpers'
import { removeCoupon } from '../remove-coupon'

describe(removeCoupon, () => {
  describe('remove from authenticated user basket', () => {
    beforeEach(() => {
      ;(axios.delete as jest.Mock).mockResolvedValue({
        data: authenticatedSprykerBasket
      })
      mget.mockResolvedValue(elasticsearchProducts)
    })

    it('creates the client & passes the locale, token, cartId, coupon to spryker request', async () => {
      await removeCoupon('en-gb', 'cart-id', 'coupon', {
        anonymousCustomerUniqueId: 'some-unique-id',
        token: 'some-user-token'
      })
      expect(createClient).toHaveBeenCalledWith({
        addAnonymousCustomerUniqueId: undefined,
        locale: 'en-gb',
        options: {
          baseURL: 'spryker.co.uk',
          headers: {
            Authorization: `Bearer some-user-token`
          }
        }
      })
      expect(axios.delete).toHaveBeenCalledWith(
        `/carts/cart-id/vouchers/coupon`
      )
    })

    it('returns the normalized basket response', async () => {
      const response = await removeCoupon('en-gb', 'cart-id', 'coupon', {
        anonymousCustomerUniqueId: 'some-unique-id',
        token: 'some-user-token'
      })
      expect(response).toMatchSnapshot()
    })
  })
  describe('remove from non-authenticated user basket', () => {
    beforeEach(() => {
      ;(axios.delete as jest.Mock).mockResolvedValue({
        data: sprykerBasket
      })
      mget.mockResolvedValue(elasticsearchProducts)
    })
    it('creates the client & passes the locale, token, cartId, coupon to spryker request', async () => {
      await removeCoupon('en-gb', 'cart-id', 'coupon', {
        anonymousCustomerUniqueId: undefined,
        token: undefined
      })
      expect(createClient).toHaveBeenCalledWith({
        addAnonymousCustomerUniqueId: undefined,
        locale: 'en-gb',
        options: {
          baseURL: 'spryker.co.uk',
          headers: undefined
        }
      })
      expect(axios.delete).toHaveBeenCalledWith(
        `/guest-carts/cart-id/vouchers/coupon`
      )
    })
  })
})
