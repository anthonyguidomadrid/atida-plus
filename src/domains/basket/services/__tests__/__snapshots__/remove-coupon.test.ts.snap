// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeCoupon remove from authenticated user basket returns the normalized basket response 1`] = `Object {}`;
